File: framework/test2/char/chars2d/Chars2d.cs
using Soda.Unittest;

public class Solution {
    public char[][] ToUpper(IList<IList<char>> matrix)
    {
        int diff = 'a' - 'A';
        int cols = matrix[0].Count;
        char[][] mx = new char[matrix.Count][];
        for (int i = 0; i < matrix.Count; ++i) {
            mx[i] = new char[cols];
            for (int j = 0; j < cols; ++j) {
                mx[i][j] = (char) (matrix[i][j] - diff);
            }
        }
        return mx;
    }
}

public class Chars2d
{
    public static void Main(string[] args)
    {
        var work = WorkFactory.Create(Utils.Fn(new Solution().ToUpper));
        // work.SetValidator((e, r) => ...);
        work.CompareSerial = true;
        Console.WriteLine(work.Run(Utils.readStdin()));
    }
}

File: framework/test2/char/chars2d/Chars2d.kt
import soda.kotlin.unittest.*

class Solution {
    fun toUpper(matrix: List<List<Char>>): Array<CharArray> {
        val diff = 'a' - 'A'
        val mx = Array(matrix.size) { CharArray(matrix[0].size) }
        for (i in matrix.indices) {
            for (j in matrix[0].indices) {
                mx[i][j] = (matrix[i][j].code.toInt() - diff).toChar()
            }
        }
        return mx
    }
}

class Chars2d : (String) -> String {
    override fun invoke(text: String): String {
        val work = GenericTestWork.create(Solution()::toUpper)

        // * by method that has not return value
        // val work = GenericTestWork.createVoid(Solution.METHOD_WITHOUT_RETURN)
        // * by class of data struct
        // val work = GenericTestWork.forStruct(STRUCT::class)

        // * setup validator
        // work.validator = (T, T) -> Boolean
        work.compareSerial = true
        return work.run(text)
    }

}

fun main() {
    print(Chars2d()(Utils.fromStdin()))
}
